src/communication/base_communicator.hpp

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+#include <random>
+#include <functional>
+
+#include <algorithms.hpp>
+#include <connection.hpp>
+#include <communication/gathered_vector.hpp>
+#include <event_queue.hpp>
+#include <spike.hpp>
+#include <util/debug.hpp>
+#include <util/double_buffer.hpp>
+#include <util/partition.hpp>
+#include <util/compare.hpp>
+
+namespace nest {
+namespace mc {
+namespace communication {
+
+// When the communicator is constructed the number of target groups and targets
+// is specified, along with a mapping between local cell id and local
+// target id.
+//
+// The user can add connections to an existing communicator object, where
+// each connection is between any global cell and any local target.
+//
+// Once all connections have been specified, the construct() method can be used
+// to build the data structures required for efficient spike communication and
+// event generation.
+template <typename CommunicationPolicy>
+class base_communicator {
```

Comment:
The classes that we derive from `communicator_base` add just one method: `make_event_queues`.
This implementation could be simplified significantly with type mixins, instead of the full
derivation.                                                                                                                                                                                                         

I can't see any heavy type dependencies, and indeed some of the types can come from the mixin type:
`base::event_queue` is only used inside the specialization, so define it there.
`base::lt_spike_source`: we only do comparisons inside the specialization, so define it there (or use a lambda)

The information you get from: `base::num_groups_local`, `base::cell_group_index`, and `base::connections_` could be passed directly. Here is some pseudo-code:

```
template <typename CommunicationPolicy, typename EventQueueImpl>
class base_communicator {
    using queue_impl = EventQueueImpl;
    queue_impl queue_impl_;
public:
    using event_queue = typename queue_impl::event_queue;

    // define make_event_queues here
    std::vector<event_queue> make_event_queues(const gathered_vector<spike>& global_spikes) {
        return queue_impl_(global_spikes_, connections_, cell_gid_partition_);
    }   
};

// then pick your specialization as follows:
template <typename CommunicationPolicy>
using communicator = communication_base<CommunicationPolicy, global_search_policy>;